cmd/swarm/swarm-smoke-pss/pss.go

```diff
@@ -0,0 +1,364 @@
+			}
+		default:
+			err = fmt.Errorf("invalid pssMode %d", mode)
+			log.Error("error processing job type", "err", err)
+		}
+
+		err = s.waitForMsg()
+		if err != nil {
+			log.Error("error while waiting for msg", "err", err)
+		}
+		if err != nil {
+			errs = append(errs, err)
+		}
+
+	}
+	if len(errs) > 0 {
+		return len(jobs) - len(errs), fmt.Errorf("%d/%d jobs failed processing", len(errs), len(jobs))
+	}
+	return len(jobs), nil
+}
+
+func (s *pssSession) genJobs(count int, mode pssMode, msgSizeBytes int) []pssJob {
+	jobs := make([]pssJob, count)
+	for i := 0; i < count; i++ {
+		jobs[i] = s.genJob(mode, msgSizeBytes)
+	}
+	return jobs
+}
+
+// genJob generates a pssJob with random message that will be sent
+// from a random sending node to a random receiving node
+func (s *pssSession) genJob(mode pssMode, msgSizeBytes int) pssJob {
+	senderNodeIdx := rand.Intn(len(s.nodes))
+	senderNode := s.nodes[senderNodeIdx]
+	log.Trace("sender node", "pss_baseAddr", hexutil.Encode(senderNode.addr), "host", hosts[senderNodeIdx])
+
+	// receiving node has to be different than sender
+	recvNodeIdx := senderNodeIdx
+	for recvNodeIdx == senderNodeIdx {
+		recvNodeIdx = rand.Intn(len(s.nodes))
+	}
+	recvNode := s.nodes[recvNodeIdx]
+	log.Trace("recv node", "pss_baseAddr", hexutil.Encode(recvNode.addr), "host", hosts[recvNodeIdx])
+
+	// create new message and add it to job index to check for receives
+	randomMsg := testutil.RandomBytes(seed, msgSizeBytes)
+	// change seed so that the next random message is different
+	seed++
+
+	j := pssJob{
+		sender:   senderNode,
+		receiver: recvNode,
+		mode:     mode,
+		msg:      randomMsg,
+	}
+
+	msgIdx := toMsgIdx(randomMsg)
+	s.jobs[msgIdx] = &j
+
+	log.Debug("gen job", "job", hexutil.Encode([]byte(msgIdx)), "sender", hosts[j.sender.hostIdx], "recv", hosts[j.receiver.hostIdx])
+
+	return j
+}
+
+// waitForMsg blocks until a msg is received or a timeout is reached
+func (s *pssSession) waitForMsg() error {
+	select {
+	case res := <-s.msgC:
+		resIdx := toMsgIdx(res.Msg)
+		resJob, ok := s.jobs[resIdx]
+		if !ok {
+			return fmt.Errorf("corrupt message for job %s", resIdx)
+		}
+		if !bytes.Equal(res.Msg, resJob.msg) {
+			return fmt.Errorf("message mismatch. expected: %s got: %s", resJob.msg, res.Msg)
+		}
+		log.Debug("got msg", "job", hexutil.Encode([]byte(resIdx)))
+	case <-time.After(time.Duration(pssMessageTimeout) * time.Second):
+		return fmt.Errorf("message timeout after %d sec", pssMessageTimeout)
+	}
+	return nil
+}
+
+func (s *pssSession) sendRawMessage(sender *pssNode, receiver *pssNode, msg []byte) error {
+	return sender.client.Call(nil, "pss_sendRaw", hexutil.Encode(receiver.addr), s.topic, hexutil.Encode(msg))
+}
+
+func (s *pssSession) sendAsymMessage(sender *pssNode, receiver *pssNode, msg []byte) error {
+
+	// share public keys between nodes
+	err := sender.client.Call(nil, "pss_setPeerPublicKey", receiver.pubkey, s.topic, hexutil.Encode(receiver.addr))
```

Comment:
Strictly speaking you don't have to do this every time you send an asym message. You only add to address book once.

Reply:
I've added a map to track the known pubkeys. Now I'm only calling `setPeerPublicKey` once.